Step2

imcool2551 opened the pull request.

imcool2551 committed:
docs: README

기능 목록 정리

imcool2551 committed:
feat+test: PinNo

볼링 한 번의 투구 결과 쓰러뜨린 핀의 개수를 나타냅니다.

imcool2551 committed:
feat+test: PinNumbers

볼링에서 총 10프레임 중 한 번의 프레임에서 2번의 투구 결과 쓰러뜨린 핀의 개수들을 나타냅니다.

imcool2551 committed:
feat+test: PinStatus

두 번의 투구결과 쓰러뜨린 핀의 개수들을 통해 한 프레임의 결과를 반환합니다.

imcool2551 committed:
feat: Frame

한 프레임을 나타내는 추상 클래스입니다. 다음 프레임을 반환하는 next() 메서드를 추상 메서드로 가집니다.

imcool2551 committed:
feat+test: NormalFrame

볼링의 1~9프레임을 나타냅니다. next() 에서 프레임 번호를 확인해서 NormalFrame, FinalFrame 중 하나를 반환합니다.

imcool2551 committed:
feat+test: NormalFrameNo

프레임 번호 1~9를 나타냅니다. 캐싱을 통해 재사용 됩니다.

imcool2551 committed:
feat+test: FinalFrame

볼링의 마지막 10프레임을 나타냅니다. next() 메서드는 호출 시 예외가 발생합니다.
10프레임은 1~9프레임과 달리 STRIKE/SPARE 일 때 한번 더 투구할 수 있기 때문에 addExtra()를 통해 조건을 검사하고 추가 투구 결과를 설정할 수 있도록 했습니다.

imcool2551 committed:
feat+test: Frames

1~9프레임과 10프레임을 통해 볼링결과를 나타냅니다.
List로 관리되는 1~9프레임의 사이즈를 관리해서 일반 프레임이 최대9개를 넘지 못하게 했습니다.
마지막 10프레임은 일반 프레임이 9개 모두 채워져야 추가할 수 있도록 했습니다.

imcool2551 committed:
feat+test: Player

참가자 이름을 나타냅니다.

imcool2551 committed:
feat: Bowling

한 참가자와 그의 볼링 게임 결과를 나타냅니다

imcool2551 committed:
feat: UI+Main

입출력 + 실행 담당 클래스들입니다.

yongdae commented on src/main/java/bowling/Main.java:
```diff
@@ -0,0 +1,37 @@
+
+public class Main {
+
+    public static void main(String[] args) {
```
미션에서 요구한 방식과 게임의 진행방식이 달라요.

<img width="1163" alt="스크린샷 2022-05-13 오전 12 52 39" src="https://user-images.githubusercontent.com/4317804/168117423-ab9bec7b-6462-49c3-a7c7-383f5fb8c688.png">

1단계에서 요구하는 볼링 미션처럼 게임이 진행 될 수 있게 해주세요. :)

<img width="1002" alt="스크린샷 2022-05-13 오전 1 09 51" src="https://user-images.githubusercontent.com/4317804/168120386-7b03bf0f-ffab-4ac4-87f0-e80afe375e38.png">

미션과 같은 방향으로 볼링 게임이 진행된다면, 설계해주신 내용이 전체적으로 변경이 될거라고 판단해
더 상세한 리뷰는 이 의견이 반영 된 이후에 진행하도록 하겠습니다.

그 외에도 설계 내용 전반에 걸쳐 getter를 사용하지 않고 미션을 진행 할 수 있게 해주시면 좋을 것 같습니다. 😄

yongdae reviewed (changes requested):
안녕하세요. 상욱님! 😄 

볼링 게임 첫 스텝을 잘 진행해주셨네요. :)
다만 현재 구현해주신 내용이 미션과 다른 부분이 있어요.

지금 리뷰를 진행하기보다는 게임 진행 방식이 수정 된 이후에
리뷰를 진행 하는게 좋을 것 같아 한가지 의견을 남겨 놓았습니다.

확인하시고 다시 리뷰 요청 부탁드리겠습니다. 🙇

imcool2551 committed:
docs: README

기능 수정에 따른 README 업데이트

imcool2551 committed:
refactor+test: PinNo

현재 추가하려는 Pin의 번호가 가능한 핀인지 확인하는 canPlus() 메서드를 추가하고 테스트했습니다

imcool2551 committed:
refactor+test: PinNumbers

첫 번째 투구와 두 번째 투구를 각각 설정할 수 있도록 변경했습니다.
원래 NormalFrame, FinalFrame에서 동시에 사용했지만 NormalFrame에서만 사용하도록 변경했습니다.

imcool2551 committed:
refactor+test: PinStatus

getStatus 메서드 이름을 plus 로 변경했습니다.
isMiss() 메서드는 FinalFrame이 다 찼는지 판별하는데 사용합니다.
isSpare() 메서드는 PinNo를 출력할 때 스페어를 구분하기 위해 사용합니다.

imcool2551 committed:
refactor: FrameNo

imcool2551 committed:
refactor: Frame

FinalFrame이 더 이상 PinNumbers를 사용하지 않기 때문에 Frame을 추상클래스 대신 인터페이스로 만들었습니다.

imcool2551 committed:
refactor+test: FinalFrame

PinNumbers + PinNo 조합 대신 List<PinNo> 로 데이터를 관리하도록 변경했습니다.
Frame인터페이스의 메서드를 FinalFrame 비즈니스 로직에 맞게 구현했습니다.

imcool2551 committed:
refactor+test: NormalFrame

FinalFrame과 마찬가지로 Frame 인터페이스의 메서드를 비즈니스 로직에 맞게 구현했습니다.
PinNumbers 필드는 그대로 사용하지만 첫 번째 투구와 두 번째 투구를 따로 설정할 수 있도록 변경했습니다.

imcool2551 committed:
refactor+test: Frames

NormalFrame, FinalFrame을 구분하지 않고 List<Frame>을 통해 하나의 리스트로 관리하도록 변경했습니다.
프레임 추가 또한 구분하지 않고 addPin()을 통해 동일하게 추가합니다.

imcool2551 committed:
refactor: Bowling

imcool2551 committed:
feat+test: PinNoPrinter

View에서 Frame을 출력할 때 Frame의 PinNo를 문자열로 변환해주는 유틸 클래스입니다.

imcool2551 committed:
refactor: UI + Main

yongdae commented on README.md:
```diff
@@ -1,9 +1,50 @@
\ No newline at end of file
+# 볼링 게임 점수판 
+
+### 기능 목록 정리
```
기능 목록을 추가해주셨네요. 👍 

다만 `기능 목록`이기보다는 `설계`에 가까운 것 같아요.
프로젝트 설계와 상관없는 볼링 점수판 기능을 목록으로 정리해서 추가해보면 어떨까요? 🤔

yongdae commented on src/test/java/bowling/domain/pin/PinNumbersTest.java:
```diff
@@ -0,0 +1,37 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class PinNumbersTest {
```
PinNumbers를 이용해서 투구에 대한 테스트를 정확하게 만들어주셨네요. 👍

yongdae commented on src/test/java/bowling/domain/pin/PinNoTest.java:
```diff
@@ -0,0 +1,47 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class PinNoTest {
```
쓰러트린 핀도 객체로 잘 만들어서 테스트도 멋지게 만들어주셨네요. 💯

yongdae commented on src/test/java/bowling/domain/frame/NormalFrameTest.java:
```diff
@@ -0,0 +1,27 @@
+    void nextFrame_WhenCalledAtMaxFrameNo_ReturnsFinalFrame() {
+        NormalFrame frame = new NormalFrame(9, 0);
+
+        assertThat(frame.nextFrame(10)).isInstanceOf(FinalFrame.class);
```
마지막 프레임도 테스트 코드로 잘 검증해주셨네요. 😄

yongdae commented on src/main/java/bowling/domain/pin/PinStatus.java:
```diff
@@ -0,0 +1,25 @@
+package bowling.domain.pin;
+
+public enum PinStatus {
```
상태를 Enum이 아니라 Class로 정의해본다면 어떨까요?

yongdae commented on src/main/java/bowling/domain/pin/PinNumbers.java:
```diff
@@ -0,0 +1,33 @@
+        this.secondNo = PinNo.of(pinNo);
+    }
+
+    public PinNo getFirstNo() {
+        return firstNo;
+    }
+
+    public PinNo getSecondNo() {
+        return secondNo;
+    }
```
`규칙 9: 게터/세터/프로퍼티를 쓰지 않는다.` 이 규칙을 적용해보면 좋을 것 같아요.

yongdae commented on src/main/java/bowling/domain/game/Bowling.java:
```diff
@@ -0,0 +1,26 @@
+
+public class Bowling {
+
+    private final Player player;
```
Player는 사용되지 않는 것 같아요.

yongdae commented on src/main/java/bowling/domain/frame/Frames.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public List<Frame> getFrames() {
+        return new ArrayList<>(frames);
```
Collections.unmodifiableList()를 사용해보면 어떨까요?

yongdae commented on src/main/java/bowling/domain/game/Bowling.java:
```diff
@@ -0,0 +1,26 @@
+
+import bowling.domain.frame.Frames;
+
+public class Bowling {
```
Bowling에 대한 테스트 코드를 추가해보면 좋을 것 같아요. :)

yongdae reviewed (approved):
볼링 점수판 그리기 첫 단계를 잘 만들어주셨네요. 
미션 난이도가 있음에도 전체적은 구조도 깔끔하네요 💯 

크게 의견 드릴 부분이 없다고 생각해 바로 다음 단계로 넘어가도록 하겠습니다!!
소소하지만 몇 가지 의견을 남겨놓은 부분은 다음 단계에서 같이 반영 부탁드릴께요. 🔥 

수고하셨습니다. 🙇
